File: monitorrent/rest/topics.py
from builtins import str
from builtins import object
import falcon
import six
from monitorrent.plugin_managers import TrackersManager


# noinspection PyUnusedLocal
class TopicCollection(object):
    def __init__(self, tracker_manager):
        """
        :type tracker_manager: TrackersManager
        """
        self.tracker_manager = tracker_manager

    def on_get(self, req, resp):
        resp.json = self.tracker_manager.get_watching_topics()

    def on_post(self, req, resp):
        body = req.json
        try:
            url = body['url']
            settings = body['settings']
            added_id = self.tracker_manager.add_topic(url, settings)
        except KeyError:
            raise falcon.HTTPBadRequest('WrongParameters', 'Can\'t add topic')
        if not added_id:
            raise falcon.HTTPInternalServerError('ServerError', 'Can\'t add topic')
        resp.set_header('Location', '/api/topics/{0}'.format(added_id))
        resp.status = falcon.HTTP_201


class TopicParse(object):
    def __init__(self, tracker_manager):
        """
        :type tracker_manager: TrackersManager
        """
        self.tracker_manager = tracker_manager

    def on_get(self, req, resp):
        url = six.text_type(req.get_param('url', required=True))
        title = self.tracker_manager.prepare_add_topic(url)
        if not title:
            raise falcon.HTTPBadRequest('CantParse', 'Can\'t parse url: \'{}\''.format(url))
        resp.json = title


# noinspection PyUnusedLocal,PyShadowingBuiltins
class Topic(object):
    def __init__(self, tracker_manager):
        """
        :type tracker_manager: TrackersManager
        """
        self.tracker_manager = tracker_manager

    def on_get(self, req, resp, id):
        try:
            topic = self.tracker_manager.get_topic(id)
        except KeyError as e:
            raise falcon.HTTPNotFound(title='NotFound', description='Can\'t find topic: {0}'.format(id))
        resp.json = topic

    def on_put(self, req, resp, id):
        settings = req.json
        try:
            updated = self.tracker_manager.update_topic(id, settings)
        except KeyError as e:
            raise falcon.HTTPNotFound(title='Id {0} not found'.format(id), description=str(e))
        if not updated:
            raise falcon.HTTPInternalServerError('ServerError', 'Can\'t update topic {}'.format(id))
        resp.status = falcon.HTTP_204

    def on_delete(self, req, resp, id):
        try:
            deleted = self.tracker_manager.remove_topic(id)
        except KeyError as e:
            raise falcon.HTTPNotFound(title='Id {0} not found'.format(id), description=str(e))
        if not deleted:
            raise falcon.HTTPInternalServerError('ServerError', 'Can\'t delete topic {}'.format(id))
        resp.status = falcon.HTTP_204


# noinspection PyUnusedLocal,PyShadowingBuiltins
class TopicResetStatus(object):
    def __init__(self, tracker_manager):
        """
        :type tracker_manager: TrackersManager
        """
        self.tracker_manager = tracker_manager

    def on_post(self, req, resp, id):
        try:
            updated = self.tracker_manager.reset_topic_status(id)
        except KeyError as e:
            raise falcon.HTTPNotFound(title='Id {0} not found'.format(id), description=str(e))
        if not updated:
            raise falcon.HTTPInternalServerError('ServerError', 'Can\'t reset topic {} status'.format(id))
        resp.status = falcon.HTTP_204


# noinspection PyUnusedLocal,PyShadowingBuiltins
class TopicPauseState(object):
    def __init__(self, tracker_manager):
        """
        :type tracker_manager: TrackersManager
        """
        self.tracker_manager = tracker_manager

    def on_post(self, req, resp, id):
        state = req.json
        paused = state.get('paused', None)
        if not isinstance(paused, bool):
            raise falcon.HTTPBadRequest('BadRequest', "'pause' has to exist and be bool")
        try:
            updated = self.tracker_manager.set_topic_paused(id, paused)
        except KeyError as e:
            raise falcon.HTTPNotFound(title='Id {0} not found'.format(id), description=str(e))
        if not updated:
            raise falcon.HTTPInternalServerError('ServerError', 'Can\'t set topic {} pause'.format(id))
        resp.status = falcon.HTTP_204

File: src/api/topics.js
import { throwOnError } from './helper'

const topicParse = /\/api\/topics\/(\d+)\/?/

export default {
    all () {
        return fetch('/api/topics')
            .then(throwOnError)
            .then(data => data.json())
    },

    get (id) {
        return fetch(`/api/topics/${id}`)
            .then(throwOnError)
            .then(response => response.json())
    },

    setPaused (id, value) {
        return fetch(`/api/topics/${id}/pause`, { method: 'POST', body: JSON.stringify({ paused: value }) })
            .then(throwOnError)
    },

    resetStatus (id) {
        return fetch(`/api/topics/${id}/reset_status`, { method: 'POST', body: '' })
            .then(throwOnError)
    },

    delete (id) {
        return fetch(`/api/topics/${id}`, { method: 'DELETE', body: '' })
            .then(throwOnError)
    },

    parseUrl (url) {
        return fetch('/api/topics/parse?url=' + encodeURIComponent(url))
            .then(throwOnError)
            .then(response => response.json())
    },

    add (url, settings) {
        return fetch('/api/topics', { method: 'POST', body: JSON.stringify({url, settings}) })
            .then(throwOnError)
            .then(response => {
                const location = response.headers.get('Location')
                const match = topicParse.exec(location)

                return parseInt(match[1])
            })
    },

    edit (id, settings) {
        return fetch(`/api/topics/${id}`, { method: 'PUT', body: JSON.stringify(settings) })
            .then(throwOnError)
    }
}
